src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp

```diff
@@ -0,0 +1,484 @@
+
+    FrameRendParams frame_rend_params;
+    frame_rend_params.background = Color(GetCOREInterface()->GetBackGround(time, FOREVER));
+    frame_rend_params.regxmin = frame_rend_params.regymin = 0;
+    frame_rend_params.regxmax = frame_rend_params.regymax = 1;
+    
+    // Collect the entities we're interested in.
+    if (m_progress_cb)
+        m_progress_cb->SetTitle(_T("Collecting Entities..."));
+
+    m_entities.clear();
+
+    MaxSceneEntityCollector collector(m_entities);
+    collector.collect(m_scene_inode);
+
+    // Call RenderBegin() on all object instances.
+    render_begin(m_entities.m_objects, time);
+
+    // Build the project.
+    if (m_progress_cb)
+        m_progress_cb->SetTitle(_T("Building Project..."));
+
+    asf::auto_release_ptr<asr::Project> project(
+        build_project(
+            m_entities,
+            m_default_lights,
+            m_view_inode,
+            view_params,
+            rend_params,
+            frame_rend_params,
+            renderer_settings,
+            m_bitmap,
+            time));
+
+    std::setlocale(LC_ALL, previous_locale.c_str());
+
+    return project;
+}
+
+void AppleseedIInteractiveRender::update_camera(INode* camera)
+{
+    ViewParams view_params;
+    get_view_params_from_view_node(view_params, camera, m_time);
+
+    m_project->get_scene()->get_active_camera()->transform_sequence().set_transform(
+        (float)m_time,
+        asf::Transformd::from_local_to_parent(to_matrix4d(Inverse(view_params.affineTM))));
+}
+
+//
+// IInteractiveRender implementation
+//
+
+void AppleseedIInteractiveRender::BeginSession()
+{
+    if (m_render_session == nullptr)
+    {
+        RendererSettings renderer_settings = RendererSettings::defaults();
+
+        m_time = GetCOREInterface()->GetTime();
+
+        ViewExp13* vp13 = reinterpret_cast<ViewExp13*>(GetViewExp()->Execute(ViewExp::kEXECUTE_GET_VIEWEXP_13));
+        INode* active_cam = vp13->GetViewCamera();
+
+        ViewParams view_params;
+        if (GetUseViewINode())
+            get_view_params_from_view_node(view_params, GetViewINode(), m_time);
+        else
+            get_view_params_from_viewport(view_params, *GetViewExp(), m_time);
+
+        m_project = prepare_project(renderer_settings, view_params, m_time);
+
+        m_render_session.reset(new InteractiveSession(
+            m_iirender_mgr, 
+            m_project.get(),
+            renderer_settings,
+            m_bitmap
+            ));
+
+        if (m_progress_cb)
+            m_progress_cb->SetTitle(_T("Rendering..."));
+
+        m_currently_rendering = true;
+
+        m_node_callback.reset(new SceneChangeCallback(this, active_cam));
+        m_callback_key = GetISceneEventManager()->RegisterCallback(m_node_callback.get(), false, 100, true);
+
+        m_render_session->start_render();
+    }
+}
+
+void AppleseedIInteractiveRender::EndSession()
+{
+    m_currently_rendering = false;
+
+    if (m_render_session != nullptr)
```

Comment:
EndSession is actually called many times (from the destructor) without any session being created, so it is possible to have m_render_session nullptr here. In BeginSession check was redundant because it is only called when renderering started.

Reply:
Got rid of m_currently_rendering. Seems to be working fine..